Lib/asyncio/tasks.py

```diff
@@ -73,15 +73,27 @@ class Task(futures._PyFuture):  # Inherit Python Task implementation
     """A coroutine wrapped in a Future."""
 
     # An important invariant maintained while a Task not done:
+    # _fut_waiter is either None or a Future.  The task can be
+    # in any of 3 states:
     #
-    # - Either _fut_waiter is None, and _step() is scheduled;
-    # - or _fut_waiter is some Future, and _step() is *not* scheduled.
+    # - 1 (_fut_waiter is not None and not _fut_waiter.done());
+    #   __step() is *not* scheduled and the Task is waiting for _fut_waiter.
+    # - 2a (_fut_waiter is None or _fut_waiter.done()), and __wakeup() is scheduled;
+    #   the Task is waiting for __wakeup() to be executed.
+    # - 2b (_fut_waiter is None or _fut_waiter.done()), and __step() is scheduled;
+    #   the Task is waiting for __step() to be executed.
+    # - 3 _fut_waiter is None and __step() is *not* scheduled;
+    #   the Task is currently executing (in __step()).
     #
-    # The only transition from the latter to the former is through
-    # _wakeup().  When _fut_waiter is not None, one of its callbacks
-    # must be _wakeup().
-
-    # If False, don't log a message if the task is destroyed whereas its
+    # The transition from 1 to 2a happens when _fut_waiter becomes done(),
+    # as it schedules __wakeup() to be called.
+    # The transition from 2a to 2b happens when __wakeup() is executed,
+    # scheduling __step() to be called, leaving _fut_waiter in place.
```

Comment:
This confuses me. Once `__wakeup()` executes, it proceeds without waiting to call `__step()` -- there's no scheduling involved that I can see in the source code (unless the .c extension is different). The call to `future.result()` immediately produces a value or raises -- it doesn't block (there's no `await` there). I do think it is possible to enter step 2b without going through 2a when `__step()` is scheduled directly by `__init__()`. So maybe it's clearer to say that we go through either 2a or 2b?

Reply:
I may have got my wires crossed here, because it is all confusing. Initially, I thought __wakeup() was called directly (futures just calling registered callbacks) and _wakeup scheduling __step,..  But then it turns out that futures _schedule_ their callbacks... I'll have another look.

Reply:
Yes, you were right.  So, there is no 2a or2b anymore just 2.
The main point I'm trying to get across here is that there _can_ be a `_fut_waiter` present in a `done` state, i.e. the
mere presence of a `_fut_waiter` does not indicate that the Task is still blocked.   I hope I'm not making this sound more complicated than it is.

Reply:
> the mere presence of a `_fut_waiter` does not indicate that the Task is still blocked

Yes, that's crucial information.